doc/analyzers/VSTHRD003.md

```diff
@@ -90,3 +92,116 @@ class SomeClass
     }
 }
 ```
+
+## Advanced cases
+
+In the next example, a `TaskCompletionSource<T>` is used as a black-box for unblocking functionality.
+It too, represents awaiting a foreign task:
+
+```cs
+class SomeClass
+{
+    TaskCompletionSource<bool> tcs = new();
+
+    public async Task MyMethodAsync()
+    {
+        await this.tcs.Task; /* This analyzer will report warning on this line. */
+        /* do more stuff */
+    }
+
+    void UnlockProgress()
+    {
+        this.tcs.TrySetResult(true);
+    }
+}
+```
+
+The problem with the above code is that `MyMethodAsync()` waits for unknown work before making progress.
+If `UnlockProgress()` is never called, the caller of `MyMethodAsync()` will be awaiting forever.
+Now suppose that the caller of `MyMethodAsync()` is actually inside a `JoinableTaskFactory.Run` delegate:
+
+```cs
+void SomeCaller()
+{
+    joinableTaskFactory.Run(async delegate
+    {
+        await someClass.MyMethodAsync();
+    });
+}
+```
+
+If `SomeCaller()` runs on the main thread, then it will effectively block the main thread while waiting for `this.tcs.Task` from `SomeClass` to complete.
+Now suppose that another thread comes along and wants to do some work before calling `UnlockProgress()`:
+
+```cs
+async Task KeyMasterAsync()
+{
+    await joinableTaskFactory.SwitchToMainThreadAsync();
+    // do some work
+    // Unblock others
+    someClass.UnlockProgress();
+}
+```
+
+We have a deadlock, because `SomeCaller()` is blocking the main thread while waiting for `UnlockProgress()` to be called, but `UnlockProgress()` will not be called until `KeyMasterAsync` can reach the main thread.
+
+Fixing this fundamentally means that `SomeCaller` will need to *join* whatever work may be needed to ultimately call `UnlockProgress`. But for `SomeCaller`, that work is unknown, since it's at least partially inside another class.
+`TaskCompletionSource<T>` is fundamentally a blackbox and the most difficult thing to use correctly while avoiding deadlocks.
+If you can instead use `JoinableTaskFactory.RunAsync` to execute your code as given in the solution above, that is ideal.
+
+Assuming you must keep using `TaskCompletionSource<T>` though, here's how it can be done as safely as possible.
```

Comment:
Would it be worth offering other options here? For one-time work, AsyncLazy with JTF integraton could be a better alternative for example.